### PR TITLE
Add Serverless section with AWS Lambda docs 

### PR DESCRIPTION
**Pull request by samimusallam**
Added a new Serverless section in `content/docs/instrumentation/js/` with AWS Lambda documentation.

One thing to note is that I used Serverless Framework for the deployment section since it makes things a lot easier, and supports other cloud vendors (can be used for Azure docs, etc.). Let me know if it makes sense to keep this in the docs.

### CODE REVIEW

**Comment by svrnm** [comment]
Preview: https://deploy-preview-1974--opentelemetry.netlify.app/docs/instrumentation/js/serverless/

**Review by svrnm — CHANGES_REQUESTED** [review]

**Comment by svrnm** [comment]
@open-telemetry/docs-approvers , in accordance with our plans to have more docs and not blog posts, I asked @samimusallam to take his blog post on lambda and turn it into a doc.

File: content/en/docs/instrumentation/js/serverless/_index.md
Comment (chalin) [inline content/en/docs/instrumentation/js/serverless/_index.md:3]: This alias should be dropped. The other similar aliases that are present in older pages are there for historical reasons. We don't create new aliases to pages to the non-existent section `/docs/js`.

File: content/en/docs/instrumentation/js/serverless/aws-lambda.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless/aws-lambda.md:98]: Will use opentelemetry.io here

File: content/en/docs/instrumentation/js/serverless/aws-lambda.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless/aws-lambda.md]: done :)

File: content/en/docs/instrumentation/js/serverless/_index.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless/_index.md]: Thanks for letting me know! dropped.

File: content/en/docs/instrumentation/js/serverless.md
Comment (svrnm) [inline content/en/docs/instrumentation/js/serverless.md:29]: I think you don't need all of them because they have some interdependencies, can you verify what the smallest set of packages is, that you really need?

File: content/en/docs/instrumentation/js/serverless.md
Comment (svrnm) [inline content/en/docs/instrumentation/js/serverless.md:105]: this looks surprising, shouldn't there be some dependency between the http.get and the return of that function? like return the statusCode of the http.get in your return object?

**Review by svrnm — CHANGES_REQUESTED** [review]
two more changes

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md:29]: Just verified that this is the smallest number of packages that I need

File: content/en/docs/instrumentation/js/serverless.md
Comment (svrnm) [inline content/en/docs/instrumentation/js/serverless.md:107]:
```suggestion
module.exports.hello = async (event) => {
  return https.get('https://opentelemetry.io/', res => {
    return {
      statusCode: res.statusCode,
    };
  }).on('error', err => {
    console.log('Error: ', err.message);
  });
};
```
Not a lambda expert, but I assume you need to return something from the handler?

File: content/en/docs/instrumentation/js/serverless.md
Comment (chalin) [inline content/en/docs/instrumentation/js/serverless.md:149]:
```suggestion
any other file: the `NODE_OPTIONS` setting ensures this.
```

**Review by chalin — APPROVED** [review]
See inline comment.

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md]: Looks a lot better, thanks! :)

**Review by svrnm — APPROVED** [review]

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:60]: Hmmm. I am not sure about this. My understanding is that batches can get lost on AWS Lambda. Don't know about other serverless environments. But I'm not an expert here either.

Comment (svrnm) [reply content/en/docs/instrumentation/js/serverless.md:60]: Probably there needs to be a [force flush](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#forceflush) somewhere?

Comment (cartermp) [reply content/en/docs/instrumentation/js/serverless.md:60]: I think that can also hurt performance in general. I'd be okay with getting this out there as a first cut but we can't have things documented that drop spans or significantly screw up performance.

Maybe @Aneurysm9 has some suggestions on what to document here?

Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md:60]: The BatchSpanProcessor does flush on [shutdown](https://github.com/open-telemetry/opentelemetry-js/blob/27c4f762948c9b8d693d5ce1c2df5aaa5b0e92ee/packages/opentelemetry-sdk-trace-base/src/export/BatchSpanProcessorBase.ts#L99), so it should send all traces before the lambda shutdown.
I'm not sure how big of a performance difference it is compared to a SimpleSpanProcessor for example, but the BatchSpanProcessor is also used in the [OpenTelemetry AWS Lambda Layer](https://github.com/open-telemetry/opentelemetry-lambda/blob/3971af6d6f47e9dc138f9d45448880e3ba8ae37f/nodejs/packages/layer/src/wrapper.ts#L101)

Comment (Aneurysm9) [reply content/en/docs/instrumentation/js/serverless.md:60]: It is generally recommended to use the `BatchSpanProcessor` everywhere but in testing where blocking export on span end can be important and tolerated.  With Lambda functions it is generally advised to call `ForceFlush` on the `(Trace|Meter)Provider` at the end of your function invocation to ensure that any telemetry is emitted before the runtime can be frozen.

There are [known issues](https://github.com/open-telemetry/opentelemetry-lambda/issues/224) with telemetry getting dropped by the Lambda layers, but they relate to extension lifecycle management issues in the collector extension layer and should not be affected by choice of span processor in the function space.

File: content/en/docs/instrumentation/js/serverless.md
Comment (Aneurysm9) [inline content/en/docs/instrumentation/js/serverless.md:82]: I believe this is only true if active tracing is disabled on your function.  When it is enabled then a sampled trace context will be available.

File: content/en/docs/instrumentation/js/serverless.md
Comment (Aneurysm9) [inline content/en/docs/instrumentation/js/serverless.md:35]: This involves reproducing a substantial portion of the OTel Lambda layers.  Should the documentation provide guidance on how to use those layers instead of directing the user to create their own wrapper?

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [reply content/en/docs/instrumentation/js/serverless.md:60]: Makes sense! Yeah, in this case ensuring there's a call at the end of the function invocation to forceFlush should then suffice.

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:10]:
```suggestion
The following show how to use OpenTelemetry to instrument AWS Lambda functions that send traces to a configured backend.
```

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:14]:
```suggestion
First, creat an empty package.json:
```

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:20]:
```suggestion
Then install the required dependencies:
```

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:90]:
```suggestion
Now that you have a Lambda wrapper, create a simple handler that servers
as a Lambda function. Save the following code as `handler.js`.
```

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:34]:
```suggestion
This file contains all the OpenTelemetry logic, which enables tracing.
```

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [reply content/en/docs/instrumentation/js/serverless.md:35]: Hmmm. Would there be any benefit to _not_ configuring and using the layer? I would say yes, we should show that, but I would want a good accounting of the pros and cons.

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:99]: Is `async` and `await` possible to use in lambda? Would prefer to show that instead of promises if possible

File: content/en/docs/instrumentation/js/serverless.md
Comment (cartermp) [inline content/en/docs/instrumentation/js/serverless.md:151]: This seems to me like we need an additional section for how to properly package and deploy the app for Lambda when not using serverles.

**Review by cartermp — CHANGES_REQUESTED** [review]
Thanks! Left a review pass and some changes that need to be made. Overall I like this content a lot.

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md]: I tried making the documentation as simple as possible to get the user started with tracing their Lambdas, letting them know the hows and the whys.
Using the layer is also possible, but it has to be compiled and deployed by the user.

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md:130]: I'll link the Serverless Framework's Getting Started guide.
Other than using AWS Console or AWS CLI, I'm not familiar with other deployments methods

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md:99]: I went with this as a reference since I want to wait for the response and return it
https://docs.aws.amazon.com/lambda/latest/dg/nodejs-handler.html#nodejs-handler-async

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md:82]: True, since by default active tracing is not enabled, I've set this option to true

File: content/en/docs/instrumentation/js/serverless.md
Comment (svrnm) [reply content/en/docs/instrumentation/js/serverless.md:130]: As long as we clarify that this is _one_ of multiple ways of doing it, I have no issue with being prescriptive in a first version of the doc. Later we still can provide options. @samimusallam besides having the links to the serverless framework maybe call out those other options without documenting them, e.g.

```suggestion
There are multiple ways of deploying your lambda function:

  * AWS Console
  * AWS CLI
  * [Serverless Framework](https://github.com/serverless/serverless)
  * ...
  
Here we will be using Serverless Framework
```

File: content/en/docs/instrumentation/js/serverless.md
Comment (svrnm) [reply content/en/docs/instrumentation/js/serverless.md]: Via the [AWS Distro docs](https://aws-otel.github.io/docs/getting-started/lambda/lambda-js) the description _"Users can enable and disable OpenTelemetry for their Lambda function without changing code."_ sounds like we have a "manual vs automatic instrumentation" situation here? 

Agreed with @cartermp we need to have a good understanding of the Pros/Cons of using the one vs the other, that makes it easier to decide if we should have both or only one of them here in the docs. 

If the answer is "both" for the sake of closing this PR we could start with an extra section that calls out the existence of the layer, would that work?

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md]: I've added a list of other options with their respective links

**Review by cartermp — APPROVED** [review]
Looks good to me as a first version of the doc.

File: content/en/docs/instrumentation/js/serverless.md
Comment (Aneurysm9) [reply content/en/docs/instrumentation/js/serverless.md:82]:
```suggestion
Note that `disableAwsContextPropagation` is set to true. The reason for this is
that the Lambda instrumentation tries to use the X-Ray context headers by
default, this results in a non-sampled context unless active tracing is enabled for this function, which creates a
`NonRecordingSpan`.
```

That should be made explicit in the text.  Readers may be confused and think it is not at all possible to get a recording span or to use the AWS context propagation carrier.

File: content/en/docs/instrumentation/js/serverless.md
Comment (Aneurysm9) [reply content/en/docs/instrumentation/js/serverless.md]: I think this documentation should at the least call out the existence of the layers and note that it duplicates a subset of their functionality.  It's good to have the DIY option spelled out, but I expect that many users are not going to want to maintain this themselves if an alternative is available.

Comment (svrnm) [reply content/en/docs/instrumentation/js/serverless.md]: > I think this documentation should at the least call out the existence of the layers and note that it duplicates a subset of their functionality. It's good to have the DIY option spelled out, but I expect that many users are not going to want to maintain this themselves if an alternative is available.

Makes sense. @samimusallam can you put a pointer to the layers for now. In a follow up we can maybe have both in the docs.

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md]: You're right, it should be noted. I pushed an update!

File: content/en/docs/instrumentation/js/serverless.md
Comment (samimusallam) [reply content/en/docs/instrumentation/js/serverless.md]: I've added a reference to it at the beginning of the docs.

**Comment by cartermp** [comment]
@Aneurysm9 any final thoughts on the latest changes?

**Review by Aneurysm9 — APPROVED** [review]
LGTM!

**Comment by cartermp** [comment]
Thanks you @samimusallam!

**Pull request merged**
